jni/hidesu.c

```diff
@@ -20,29 +22,266 @@ typedef unsigned short int sa_family_t;
+	}
+
+	if((sd = opendir("/sbin")) == NULL) {
+		umount2("/dev/sbin.tmp", MNT_DETACH);
+		return 2;
+	}
+	while((dir = readdir(sd)) != NULL) {
+		// copy files (with symlink support), copy mode/owner + SELinux context
+		strncpy(sfile+6, dir->d_name, sizeof(sfile)-6);
+		strncpy(dfile+14, dir->d_name, sizeof(dfile)-14);
+		switch(dir->d_type) {
+		case DT_REG:
+			sfd = open(sfile, O_RDONLY);
+			if(sfd == -1 || fstat(sfd, &sb) == -1) {
+				umount2("/dev/sbin.tmp", MNT_DETACH);
+				return 3;
+			}
+			if((dfd = open(dfile, O_WRONLY | O_CREAT | O_EXCL, sb.st_mode)) == -1) {
+				umount2("/dev/sbin.tmp", MNT_DETACH);
+				return 3;
+			}
+			copied = 0;
+			while((bytes = sendfile(dfd, sfd, NULL, (size_t) sb.st_size)) > 0) copied += bytes;
+			if(copied != (size_t) sb.st_size) {
+				umount2("/dev/sbin.tmp", MNT_DETACH);
+				return 3;
+			}
+			fchown(dfd, sb.st_uid, sb.st_gid);
+			fchmod(dfd, sb.st_mode);
+			res = fgetxattr(sfd, XATTR_NAME_SELINUX, buffer, sizeof(buffer) - 1);
+			if(res != -1) {
+				buffer[res] = 0;
+				res = fsetxattr(dfd, XATTR_NAME_SELINUX, buffer, strlen(buffer) + 1, 0);
+			}
+			if(res == -1) {
+				fsetxattr(dfd, XATTR_NAME_SELINUX, DEFAULT_CONTEXT, strlen(DEFAULT_CONTEXT) + 1, 0);
+			}
+			close(sfd);
+			close(dfd);
+			break;
+		case DT_LNK:
+			if((res = readlink(sfile, buffer, sizeof(buffer)-1)) == -1) {
+				umount2("/dev/sbin.tmp", MNT_DETACH);
+				return 3;
+			}
+			buffer[res] = 0;
+			if(symlink(buffer, dfile) == -1) {
+				umount2("/dev/sbin.tmp", MNT_DETACH);
+				return 3;
+			}
+			if(lstat(sfile, &sb) == 0) lchown(dfile, sb.st_uid, sb.st_gid);
+			res = lgetxattr(sfile, XATTR_NAME_SELINUX, buffer, sizeof(buffer) - 1);
+			if(res != -1) {
+				buffer[res] = 0;
+				res = lsetxattr(dfile, XATTR_NAME_SELINUX, buffer, strlen(buffer) + 1, 0);
+			}
+			if(res == -1) {
+				lsetxattr(dfile, XATTR_NAME_SELINUX, DEFAULT_CONTEXT, strlen(DEFAULT_CONTEXT) + 1, 0);
+			}
+			break;
+		}
+	}
+	closedir(sd);
+	mount("tmpfs", "/dev/sbin.tmp", "tmpfs", MS_REMOUNT | MS_RDONLY, "size=64m,nr_inodes=8192,mode=755,uid=0,gid=0");
+	if(mount("/dev/sbin.tmp", "/sbin", NULL, MS_BIND, NULL) == -1) {
+		umount2("/dev/sbin.tmp", MNT_DETACH);
+		return 4;
+	}
+	umount2("/dev/sbin.tmp", MNT_DETACH);
+	rmdir("/dev/sbin.tmp");
+
 	return 0;
 }
 
+unsigned long int getPidOfProc(char *cmd) {
+	DIR *sd;
+	struct dirent *dir;
+	char buffer[256];
+	int fd, len, cmdlen;
+	unsigned long int ret = 0;
+
+	if((sd = opendir("/proc")) == NULL) {
+		return 0;
+	}
+	cmdlen = strlen(cmd);
+	while((dir = readdir(sd)) != NULL) {
+		if(dir->d_name[0]<'0' || dir->d_name[0]>'9') continue;
+		snprintf(buffer, sizeof(buffer), "/proc/%s/cmdline", dir->d_name);
+		fd = open(buffer, O_RDONLY);
+		if(fd == -1) continue;
+		len = read(fd, buffer, sizeof(buffer)-1);
+		close(fd);
+		if(len == -1) continue;
+		buffer[len] = 0;
+		if(len>cmdlen && buffer[cmdlen]==32) buffer[cmdlen]= 0;
```

Comment:
I have made a newcommit 2a3e4e1fa3a16d17d7a51a105f9f1a7a8f4d1d99 where 32 is replaced by ' '.